client/client.go

```diff
@@ -98,6 +98,15 @@ type (
 		//	- InternalServiceError
 		StartWorkflow(ctx context.Context, options StartWorkflowOptions, workflowFunc interface{}, args ...interface{}) (*workflow.Execution, error)
 
+		// StartWorkflowAsync behaves like StartWorkflow except that the request is first queued and then processed asynchronously.
+		// See StartWorkflow for parameter details.
+		// The returned AsyncWorkflowExecution doesn't contain run ID, because the workflow hasn't started yet.
+		// The errors it can return:
+		//	- EntityNotExistsError, if domain does not exists
+		//	- BadRequestError
+		//	- InternalServiceError
```

Comment:
Are we returning InternalServiceError in thrift? This should not be included in this set since it's an unexpected error (it can happen, but there's no valid case where this should be sent to a user)

Reply:
Not sure what these errors in the comment mean to you but it says "The errors it can return" to hint the caller's err handling. All of [these](https://github.com/uber-go/cadence-client/blob/a73fe0e8ea2f180446d828ba5786a30b0668123f/internal/compatibility/thrift/error.go#L34) are possible technically This comment is doomed to be out-dated though. I just kept it similar to corresponding sync API method so they are at least consistently incomplete :)

Reply:
Yeah, I was referring to the comment. 

This is a long-running thrift thing where thift's somewhat surprising way of requiring upfront declaration of expected errors meant that people declared all errors as expected (equivalent to putting all http error codes into the 4XX range). 

Why should we care? The reason is that - a while back anyway, this messed up Muttley's ability to differentiate caller errors from callee errors (at least in thrift). So, among other things, SLAs got incorrectly higher uptime due to returning 5XX errors like 4XX errors. 

Anyway, I am probably just droning on at this point. The takeaway is that if we have any thrift APIs, we shouldn't include 5XX class errors there in the expected section (of which InternalError is one)

Reply:
IIUC you are saying all "expected" errors should fall into 4xx range which is how it should be. But that doesn't change the fact that there will be 500s (unexpected cases) and it's possible these functions may return InternalServiceError. 
I'd support removing these partial list of error types from the client code comment to avoid confusions until we revisit error propagation/handling and document it better for users.